modules/gapi/test/streaming/gapi_streaming_vpp_preproc_test.cpp

```diff
@@ -0,0 +1,491 @@
+
+    // create VPP preproc engine
+    VPPPreprocEngine preproc_engine(std::unique_ptr<VPLAccelerationPolicy>{
+                                                new VPLDX11AccelerationPolicy(device_selector)});
+
+    // launch threads
+    SafeQueue queue;
+    size_t decoded_number = 1;
+    size_t preproc_number = 0;
+
+    std::thread decode_thread([&decode_engine, sess_ptr,
+                               &queue, &decoded_number] () {
+        // decode first frame
+        {
+            cv::MediaFrame decoded_frame;
+            ASSERT_NO_THROW(decoded_frame = extract_decoded_frame(sess_ptr->session, decode_engine));
+            queue.push(std::move(decoded_frame));
+        }
+
+        // launch pipeline
+        try {
+            while(true) {
+                queue.push(extract_decoded_frame(sess_ptr->session, decode_engine));
+                decoded_number++;
+            }
+        } catch (...) {}
+
+        // send stop
+        queue.push_stop();
+    });
+
+    std::thread preproc_thread([&preproc_engine, &queue, &preproc_number, required_frame_param] () {
+        // create preproc session based on frame description & network info
+        cv::MediaFrame decoded_frame = queue.pop();
+        cv::util::optional<pp_params> first_pp_params = preproc_engine.is_applicable(decoded_frame);
+        ASSERT_TRUE(first_pp_params.has_value());
+        pp_session first_pp_sess =
+                    preproc_engine.initialize_preproc(first_pp_params.value(), required_frame_param);
+
+        // make preproc using incoming decoded frame & preproc session
+        cv::MediaFrame pp_frame = preproc_engine.run_sync(first_pp_sess, decoded_frame);
+        cv::GFrameDesc first_outcome_pp_desc = pp_frame.desc();
+
+        // launch pipeline
+        bool in_progress = false;
+        preproc_number = 1;
+        try {
+            while(true) {
+                decoded_frame = queue.pop();
+                if (SafeQueue::is_stop(decoded_frame)) {
+                    break;
+                }
+                in_progress = true;
+
+                cv::util::optional<pp_params> params = preproc_engine.is_applicable(decoded_frame);
+                ASSERT_TRUE(params.has_value());
+                ASSERT_TRUE(0 == memcmp(&params.value(), &first_pp_params.value(), sizeof(pp_params)));
+
+                pp_session pp_sess = preproc_engine.initialize_preproc(params.value(),
+                                                                       required_frame_param);
+                ASSERT_EQ(pp_sess.get<EngineSession>().get(),
+                          first_pp_sess.get<EngineSession>().get());
+
+                pp_frame = preproc_engine.run_sync(pp_sess, decoded_frame);
+                cv::GFrameDesc pp_desc = pp_frame.desc();
+                ASSERT_TRUE(pp_desc == first_outcome_pp_desc);
+                in_progress = false;
+                decoded_frame = cv::MediaFrame();
```

Comment:
good point.
Frame must be destroyed (and reused) here to do not block pipeline when decoder is waiting for a free surface to decode rather than preproc is finishing.

I faced with deadlock situation when DECODER stucks in waiting for the next free surface in decode thread but current thread stucks in `decoded_frame = queue.pop();`. Because  `decoded_frame` will be not destroyed until something extracted from `queue` and it holds a surface busy and blocks the decoder thread.

I'll try to use local variable in the loop, but in this way i still need in manually destruction the first one `decoded_frame` which acquired before loop starting ( or adding new scope {} for that)